### PR TITLE
Ds 179 picture hash

### PR DESCRIPTION
**Pull request by Knud-Aage**
Initial version of a webservice that gives a hash value of a picture

### CODE REVIEW

**Review by tokee — CHANGES_REQUESTED** [review]
The service has separate calls for the separate functions (`imageDHash`, `imagePHash`, `imageCHash`). While this is easy to understand & use, it also means a lot of calls as an indexing service will call all the methods after each other. If this structure is to be used, then the service should cache recently retrieved images, to avoid overloading the image server. Alternatively (may favourite), there should be a single `analyze`-call, where it is possible to provide a list of the calculations to perform: `http://.../analyze?methods=dhash,phash?imgURL=...`.

File: ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  - url:  /ds-pictureHash/api
+paths:
+  # Displays the DHash value given an URL like
+  #   http://localhost:8080/ds-pictureHash/api/imageDHash/imgPath/?imgPath=<path to the image>
+  /imageDHash/imgPath:
+    get:
+      summary: 'Request imageDPHash from an image at the Royal Danish Library'
+      operationId: getImageDHash
+      parameters:
+        - name: imgPath
```
Comment (tokee) [inline ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml]: This should be imgURL

File: ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  version: 1.0.0
+  title: Picture Hash
+  license:
+    name: Apache 2.0
+servers:
+  - url:  /ds-pictureHash/api
+paths:
+  # Displays the DHash value given an URL like
+  #   http://localhost:8080/ds-pictureHash/api/imageDHash/imgPath/?imgPath=<path to the image>
+  /imageDHash/imgPath:
```
Comment (tokee) [inline ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml]: It seems redundant to have `imgPath` here when it is also a parameter?

File: README.md [hunk 3]
```diff
@@ -1,3 +1,16 @@
...
+# Purpose
+
+Webservice that gives an image hash value, where the input parameter is the path to the image. Possible hash values are 
+either Perceptive Hash (PHash) or Difference Hash (DHash). 
+The PHash can be accessed by 
+http://localhost:8080/ds-pictureHash/api/imagePHash/imgPath/?imgPath=<filepath> 
+The DHash can be accessed by 
+http://localhost:8080/ds-pictureHash/api/imageDHash/imgPath/?imgPath=<filepath> 
+
+
```
Comment (tokee) [inline README.md]: Could you add a note about building & running the service?

File: conf/localtomcat9/ds-pictureHash.xml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<?xml version='1.0' encoding='utf-8'?>
+<Context override="true" docBase="ds-pictureHash.war">
+
+    <Environment name="ds-pictureHash-logback-config"
+        value="${user.home}/Code/Repository/Digitale_Samlinger/ds-pictureHash/conf/localtomcat9_halh/logback.xml"
```
Comment (tokee) [inline conf/localtomcat9/ds-pictureHash.xml]: This is a local path. Should the `localtomcat9`-folder be in the repository at all?

File: ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+components:
+  schemas:
+    HashReply:
+      type: object
+      required:
+        - message
+      properties:
+        message:
+          type: string
+          description: Hash value from the API
```
Comment (tokee) [inline ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml]: Please describe how the hash value looks (hex/int/binary/string? length?). Ideally provide an example.

File: ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  license:
+    name: Apache 2.0
+servers:
+  - url:  /ds-pictureHash/api
+paths:
+  # Displays the DHash value given an URL like
+  #   http://localhost:8080/ds-pictureHash/api/imageDHash/imgPath/?imgPath=<path to the image>
+  /imageDHash/imgPath:
+    get:
+      summary: 'Request imageDPHash from an image at the Royal Danish Library'
```
Comment (tokee) [inline ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml]: Please describe what a DPHash is or provide a link to a webpage with that description

File: ds-pictureHash-ws/src/main/java/dk/kb/api/webservice/ApiServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import java.net.URL;
+
+public class ApiServiceImpl  implements DefaultApi {
+    /**
+     * Returns the discrete hash value of the image in imgPath
+     * @param imgPath The path to the image given to the webservice.
+     * @return The discrete hash value
+     */
+    @Override
+    public HashReplyDto getImageDHash(String imgPath) {
```
Comment (tokee) [inline ds-pictureHash-ws/src/main/java/dk/kb/api/webservice/ApiServiceImpl.java]: There does not seem to be any test methods for any of the methods? Please provide at least one.

**Review by tokee — CHANGES_REQUESTED** [review]

File: ds-pictureHash-ws/src/main/java/dk/kb/api/webservice/ApiServiceImpl.java (hunk 7 above)
Comment (Knud-Aage) [reply ds-pictureHash-ws/src/main/java/dk/kb/api/webservice/ApiServiceImpl.java]: Test methods added

File: ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml (hunk 6 above)
Comment (Knud-Aage) [reply ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml]: Oops, it is now changed to DHash

File: ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml (hunk 5 above)
Comment (Knud-Aage) [reply ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml]: The format of the hash value is added - it is a BigInteger that is returned.

File: conf/localtomcat9/ds-pictureHash.xml (hunk 4 above)
Comment (Knud-Aage) [reply conf/localtomcat9/ds-pictureHash.xml]: The file is now removed from the project

File: README.md (hunk 3 above)
Comment (Knud-Aage) [reply README.md]: Explanation is now added to the README.md

File: ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml (hunk 2 above)
Comment (Knud-Aage) [reply ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml]: It apparently is the way that is possible with swagger

File: ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml (hunk 1 above)
Comment (Knud-Aage) [reply ds-pictureHash-swagger/src/main/swagger/ds-pictureHash.yml]: Done

**Pull request merged**